R/fetch_interpro.R

```diff
@@ -0,0 +1,344 @@
+                location %>%
+                  dplyr::mutate(
+                    source_database = .x[["source_database"]],
+                    source_accession = .x[["accession"]],
+                    source_name = .x[["name"]]
+                  )
+              }
+            ) %>%
+              mutate(accession = .x)
+          }
+        }
+
+        result
+      },
+      .progress = progress_option_residue
+    )
+
+    if ("issue" %in% colnames(query_residue)) {
+      problematic_ids <- query_residue %>%
+        dplyr::filter(!is.na(issue))
+
+      message("\nThere was an issue fetching from InterPro:\n", paste0(problematic_ids$accession, ": ", problematic_ids$issue, "\n"))
+      return(NULL)
+    }
+
+    return(query_residue)
+  }
+
+  # start with unnesting and flattening the data
+  flattened_query_result <- purrr::map(
+    .x = query_result,
+    .f = ~ {
+      # page level
+      purrr::map(
+        .x = .x,
+        .f = ~ {
+          .x[["results"]]
+        }
+      )
+    }
+  ) %>%
+    purrr::flatten() %>%
+    purrr::flatten()
+
+  # unnest domain level data
+  domain_result <- purrr::map_dfr(
+    .x = flattened_query_result,
+    .f = ~ {
+      go_terms <- purrr::map_dfr(
+        .x = .x[["metadata"]][["go_terms"]],
+        .f = ~ {
+          data.frame(
+            go_id = .x[["identifier"]],
+            go_name = .x[["name"]],
+            go_code = .x[["category"]][["code"]],
+            go_type = .x[["category"]][["name"]]
+          )
+        }
+      )
+
+      metadata <- data.frame(
+        identifier = .x[["metadata"]][["accession"]],
+        identifier_name = .x[["metadata"]][["name"]],
+        identifier_source_database = .x[["metadata"]][["source_database"]],
+        identifier_type = .x[["metadata"]][["type"]]
+      )
+
+      if (nrow(go_terms) != 0) {
+        metadata <- metadata %>%
+          bind_cols(go_terms)
+      }
+
+      # Protein level
+      protein_level <- purrr::map_dfr(
+        .x = .x[["proteins"]],
+        .f = ~ {
+          # Entry protein location level
+          entry_protein_locations_level <- purrr::map_dfr(
+            .x = .x[["entry_protein_locations"]],
+            .f = ~ {
+              # Fragment Level
+              protein_info <- purrr::map_dfr(
+                .x = .x[["fragments"]],
+                .f = ~ {
+                  data.frame(
+                    start = .x[["start"]],
+                    end = .x[["end"]],
+                    dc_status = .x[["dc-status"]],
+                    representative = .x[["representative"]]
+                  )
+                }
+              ) %>%
+                dplyr::mutate(
+                  model = .x[["model"]],
+                  score = .x[["score"]]
+                )
+            }
+          ) %>%
+            dplyr::mutate(accession = stringr::str_to_upper(.x[["accession"]]))
+        }
+      )
+
+      tidyr::crossing(metadata, protein_level)
+    }
+  )
+
+  domain_result
```

Comment:
i would use return(domain_result) to make the code clearer and its also used in return(query_residue)